mobify/progressive-web-scaffold
# Add Grid Component

Add `Grid` component with examples of how to get it working in a template.

 **JIRA**: [WEB-946](https://mobify.atlassian.net/browse/WEB-946)

## Changes
- Add `Grid` component
- Add `u-center-piece` utility class, who's behavior is to be full screen width until 1200px, after which it's a center aligned column.

## How to test-drive this PR
- Run `npm run styleguide`
- Visit the [Grid component](http://localhost:4000/#Grid)
- Confirm that the grid behaves as expected
- If you desire, experiment with the grid by customizing the styles in `/app/components/_example.scss`, referencing the [Susy toolkit](http://susydocs.oddbird.net/en/latest/toolkit/) for all available grid mixins.

## Proposed Grid Direction
So an important question for the scaffold we must ask is: how should we best use grids in the scaffold? Below I will outline the nature of how this PR proposes we use grids, and some comparison pro/cons with other approaches we could take.

See the comment threads below for details on each approach...

* [Grids via Wrapper Component + Template Styles](https://github.com/mobify/progressive-web-scaffold/pull/194#issuecomment-264583289) ~~(the current implementation)~~
* [Grids via Pure Template Styles](https://github.com/mobify/progressive-web-scaffold/pull/194#issuecomment-264584579)
* [Grids via Pure React Component](https://github.com/mobify/progressive-web-scaffold/pull/194#issuecomment-264587923) (the current implementation)
* [Grids via Pure Utility Classes](https://github.com/mobify/progressive-web-scaffold/pull/194#issuecomment-264589726)

All the below example grids produce the following layout, and viewable in the project Style Guide as well:

**"Mobile" breakpoint**
![screen shot 2016-12-06 at 1 11 55 pm](https://cloud.githubusercontent.com/assets/734535/20943746/a211831a-bbb5-11e6-8474-816832566054.png)

**"Tablet" breakpoint**
![screen shot 2016-12-06 at 1 11 44 pm](https://cloud.githubusercontent.com/assets/734535/20943754/a6a6fc34-bbb5-11e6-94be-5cad86c3ab46.png)

**"Desktop" breakpoint**
![screen shot 2016-12-02 at 2 45 37 pm](https://cloud.githubusercontent.com/assets/734535/20853014/072d04d8-b89e-11e6-807f-6cfe2e53ef1e.png)

## Notes
To see how I audited the available grid systems, you can view the spreadsheet [here](https://docs.google.com/spreadsheets/d/1HYW0UCmqpoEZiixfxxM-MuxOanaveZNWsjUwGhJlo-Y/edit#gid=0).

Requirements for auditing grid systems include:
* Must have a library of Sass mixins
* Must be easily customizable with custom parameters
* Parameters must be responsive across multiple media query breakpoints
* Must **not** rely on flexbox unless it has a float fallback
* React components are a nice-to-have
* Active community/maintainers is strongly preferred
